pandas/_libs/hashtable_class_helper.pxi.in

```diff
@@ -300,16 +300,19 @@ def get_dispatch(dtypes):
 
     unique_template = unique_template.format(name=name, dtype=dtype, null_condition=null_condition, float_group=float_group)
 
-    yield (name, dtype, null_condition, float_group, unique_template)
+    yield (name, dtype, null_condition, float_group, default_na_value, unique_template)
 }}
 
 
-{{for name, dtype, null_condition, float_group, unique_template in get_dispatch(dtypes)}}
+{{for name, dtype, null_condition, float_group, default_na_value, unique_template in get_dispatch(dtypes)}}
 
 cdef class {{name}}HashTable(HashTable):
 
-    def __cinit__(self, size_hint=1):
+    def __cinit__(self, size_hint=1, {{dtype}}_t na_value={{default_na_value}},
+                  bint use_na_value=False):
         self.table = kh_init_{{dtype}}()
+        self.na_value = na_value
+        self.use_na_value = use_na_value
```

Comment:
Do you even need `use_na_value` at all? Conceptually can't you just rely on the existing `check_null` implementation and simply use whatever `na_value` gets set to (whether default or provided by user)?

Reply:
I'm not sure :) Without `use_na_value`, wouldn't that break

```
pd.factorize(np.array([0, 1], dtype='u8'))
```

On master the labels `[0, 1]`. But without `use_na_value`, we'd see that `0` is equal to the default UInt64HashTable `na_value` of `0`, so the labels would be `[-1, 0]`.

I'll try that now to make sure.

Reply:
Hmm OK I see. I'll say this with the caveat that I haven't used factorize before, but does it even make sense to have a NA value for boolean types? Wouldn't the presence of NA force that conversion to float?

Reply:
I think the best use-case to focus on is Categorical. In its codes, missing values are -1. So we'd like to pass an array of int64 to factorize, and have it think that -1 is NA.

We avoid NaN for datetime / categorical for this reason: it would force them to float. For the Int64HashTable, we use iNaT (the smallest int) as the missing value sentinel.

Reply:
For Int64 or UInt64? If it's UInt64 it'll have to be 0, since -1 is negative :)

I need to better describe `default_na_value`. It's never actually used, since the only way it *could* be used is if the user explicitly passes `na_value`. Essentially, I want

```python
cdef f(int64_t x=None):
    if x is not None:
        ...
```

But in a way that makes cython happy. There's probably a way to do that...

Reply:
Good point on the -1 with the unsigned int - you'll have to excuse me (it's Friday).

So perhaps one thing to consider is instead of declaring x to be `int64_t` in your example (or whatever type is dictated by the template) you could keep it as an `object` in the signature. In the `cdef` block add the declaration for your variable, and then do your `if x is not None` construct in the body before the `nogil`, where within that conditional you could cast the object to the particular type and assign it to the variable you declared in the `cdef` section.

Kind of wordy so let me know if that's not clear

Reply:
Something like

```diff
diff --git a/pandas/_libs/hashtable_class_helper.pxi.in b/pandas/_libs/hashtable_class_helper.pxi.in
index af60eb4b1..07bdad241 100644
--- a/pandas/_libs/hashtable_class_helper.pxi.in
+++ b/pandas/_libs/hashtable_class_helper.pxi.in
@@ -409,26 +409,32 @@ cdef class {{name}}HashTable(HashTable):
     def get_labels(self, {{dtype}}_t[:] values, {{name}}Vector uniques,
                    Py_ssize_t count_prior, Py_ssize_t na_sentinel,
                    bint check_null=True,
-                   {{dtype}}_t na_value={{default_na_value}},
-                   bint use_na_value=False):
+                   object na_value=None):
         cdef:
             Py_ssize_t i, n = len(values)
             int64_t[:] labels
             Py_ssize_t idx, count = count_prior
             int ret = 0
-            {{dtype}}_t val
+            {{dtype}}_t val, na_value2
             khiter_t k
             {{name}}VectorData *ud
+            bint use_na_value
 
         labels = np.empty(n, dtype=np.int64)
         ud = uniques.data
+        use_na_value = na_value is not None
+
+        if use_na_value:
+            na_value2 = <{{dtype}}_t>na_value
+        else:
+            na_value2 = {{default_na_value}}
 
         with nogil:
             for i in range(n):
                 val = values[i]
 
                 if ((check_null and {{null_condition}}) or
-                        (use_na_value and val == na_value)):
+                        (use_na_value and val == na_value2)):
                     labels[i] = na_sentinel
                     continue
 
diff --git a/pandas/core/algorithms.py b/pandas/core/algorithms.py
index bbf60c2ee..e668b4748 100644
--- a/pandas/core/algorithms.py
+++ b/pandas/core/algorithms.py
@@ -456,16 +456,9 @@ def _factorize_array(values, check_nulls, na_sentinel=-1, size_hint=None,
     """
     (hash_klass, vec_klass), values = _get_data_algo(values, _hashtables)
 
-    use_na_value = na_value is not None
-    kwargs = dict(use_na_value=use_na_value)
-
-    if use_na_value:
-        kwargs['na_value'] = na_value
-
     table = hash_klass(size_hint or len(values))
     uniques = vec_klass()
-    labels = table.get_labels(values, uniques, 0, na_sentinel, check_nulls,
-                              **kwargs)
+    labels = table.get_labels(values, uniques, 0, na_sentinel, check_nulls, na_value=na_value)
 
     labels = _ensure_platform_int(labels)
     uniques = uniques.to_array()

```

seems to work. Is that what you had in mind?

Reply:
Just to be a little more explicit something along the lines of

```python
cdef f(object x=None):
    cdef:
        {{template_type}} x_val

    if x is not None:
        x_val = <{{template_type}}> x
    ...
    with nogil:
         ...
```

Should still work because you aren't trying to access any Python objects within the nogil - the compiler should know that x_val is of type `{{template_type}}` and work with that accordingly, even though that value doesn't get defined until runtime

Reply:
Replied at about the same time but yes that's what I was thinking. Good to hear it works